File: filelist/src/main/scala/farjs/filelist/popups/DeleteController.scala
package farjs.filelist.popups

import farjs.filelist.{FileListState, FileListUiData}
import farjs.ui.popup.{MessageBox, MessageBoxAction, MessageBoxProps}
import farjs.ui.theme.Theme
import scommons.react._

import scala.scalajs.js

object DeleteController extends FunctionComponent[FileListUiData] {

  private[popups] var messageBoxComp: ReactClass = MessageBox

  protected def render(compProps: Props): ReactElement = {
    val props = compProps.plain
    val theme = Theme.useTheme().popup

    props.data.toOption match {
      case Some(data) if props.showDeletePopup.getOrElse(false) =>
        <(messageBoxComp)(^.plain := MessageBoxProps(
          title = "Delete",
          message = "Do you really want to delete selected item(s)?",
          actions = js.Array(
            MessageBoxAction.YES { () =>
              val items =
                if (FileListState.selectedItems(data.state).nonEmpty) FileListState.selectedItems(data.state)
                else js.Array(FileListState.currentItem(data.state).toList: _*)

              props.onClose()
              data.dispatch(data.actions.deleteItems(
                dispatch = data.dispatch,
                parent = data.state.currDir.path,
                items = items
              ))
            },
            MessageBoxAction.NO { () =>
              props.onClose()
            }
          ),
          style = theme.error
        ))()
      case _ => null
    }
  }
}

File: filelist/popups/DeleteController.mjs
/**
 * @typedef {import("../FileListUi.mjs").FileListUiData} FileListUiData
 */
import React from "react";
import MessageBox from "@farjs/ui/popup/MessageBox.mjs";
import MessageBoxAction from "@farjs/ui/popup/MessageBoxAction.mjs";
import Theme from "@farjs/ui/theme/Theme.mjs";
import FileListState from "@farjs/filelist/FileListState.mjs";

const h = React.createElement;

/**
 * @param {FileListUiData} props
 */
const DeleteController = (props) => {
  const { messageBoxComp } = DeleteController;

  const theme = Theme.useTheme().popup;
  const data = props.data;

  if (data && props.showDeletePopup) {
    const onAction = () => {
      const currItem = FileListState.currentItem(data.state);
      const items = FileListState.selectedItems(data.state);
      if (items.length === 0 && currItem) {
        items.push(currItem);
      }

      props.onClose();
      data.dispatch(
        data.actions.deleteItems(data.dispatch, data.state.currDir.path, items)
      );
    };

    return h(messageBoxComp, {
      title: "Delete",
      message: "Do you really want to delete selected item(s)?",
      actions: [
        MessageBoxAction.YES(onAction),
        MessageBoxAction.NO(() => {
          props.onClose();
        }),
      ],
      style: theme.error,
    });
  }

  return null;
};

DeleteController.displayName = "DeleteController";
DeleteController.messageBoxComp = MessageBox;

export default DeleteController;
